csrc/python_frontend/fusion_record_serde.cpp

```diff
@@ -0,0 +1,901 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ops/arith.h>
+#include <ops/composite.h>
+#include <python_frontend/fusion_record_serde.h>
+#include <serde/python_fusion_cache_generated.h>
+#include <functional>
+
+namespace nvfuser::python_frontend {
+
+std::vector<State> parseStateArgs(
+    const flatbuffers::Vector<const nvfuser::serde::State*>* args) {
+  std::vector<State> result;
+  for (auto s : *args) {
+    result.emplace_back(s->index(), s->type());
+  }
+  return result;
+}
+
+template <typename T>
+std::vector<T> parseVector(const flatbuffers::Vector<T>* fb_vector) {
+  std::vector<T> result(fb_vector->begin(), fb_vector->end());
+  return result;
+}
+
+// Flatbuffer stores bool values as uint8_t.
+std::vector<bool> parseBoolVector(
+    const flatbuffers::Vector<uint8_t>* fb_vector) {
+  std::vector<bool> result(fb_vector->begin(), fb_vector->end());
+  return result;
+}
+
+static PrimDataType mapToNvfuserDtype(serde::DataType t) {
+  switch (t) {
+    case serde::DataType_Bool:
+      return PrimDataType::Bool;
+    case serde::DataType_Double:
+      return PrimDataType::Double;
+    case serde::DataType_Float:
+      return PrimDataType::Float;
+    case serde::DataType_Half:
+      return PrimDataType::Half;
+    case serde::DataType_BFloat16:
+      return PrimDataType::BFloat16;
+    case serde::DataType_Int:
+      return PrimDataType::Int;
+    case serde::DataType_Int32:
+      return PrimDataType::Int32;
+    case serde::DataType_ComplexFloat:
+      return PrimDataType::ComplexFloat;
+    case serde::DataType_ComplexDouble:
+      return PrimDataType::ComplexDouble;
+    case serde::DataType_None:
+      return PrimDataType::Null;
+    default:
+      break;
+  }
+  TORCH_INTERNAL_ASSERT(false, "No nvfuser dtype found for serde data type.");
+  return PrimDataType::Null;
+}
+
+// START deserialize functions
+
+RecordFunctor* deserializeStartRecord(const serde::RecordFunctor* buffer) {
+  return new StartRecord();
+}
+
+RecordFunctor* deserializeEndRecord(const serde::RecordFunctor* buffer) {
+  return new EndRecord();
+}
+
+template <class fn_type, class... Signature>
+RecordFunctor* deserializeOpRecord(
+    const std::unordered_map<std::string, fn_type>& str_to_func_map,
+    serde::RecordType record_type,
+    const serde::RecordFunctor* buffer) {
+  return new OpRecord<Signature...>(
+      parseStateArgs(buffer->args()),
+      parseStateArgs(buffer->outputs()),
+      buffer->name()->str(),
+      record_type,
+      str_to_func_map.at(buffer->name()->str()));
+}
+
+RecordFunctor* deserializeBatchNormRecord(const serde::RecordFunctor* buffer) {
```

Comment:
nitpick: let's hide these inside `namespace { }`.

Reply:
On a second thought, what do you say if we put this inside a lambda, instead of defining every thing as a function here?